File: packages/honker-node/test/phase_mantle.js
// Tests for Phase Mantle: schedule lifecycle (pause/resume/list/update)
// and queue cancel/getJob.

'use strict';

const test = require('node:test');
const assert = require('node:assert/strict');

const honker = require('..');
const { createTempDb } = require('./helpers');

function tmpdb() {
  return createTempDb('honker-mantle-', honker.open.bind(honker));
}

// ---------- schedule lifecycle ----------

test('schedule list round-trips all fields', () => {
  const { path: dbPath, open, cleanup } = tmpdb();
  let db;
  try {
    db = open(dbPath);
    const sched = new honker.Scheduler(db);
    sched.add({ name: 'daily-recap', queue: 'emails', cron: '0 9 * * *', payload: { foo: 1 }, priority: 3 });
    sched.add({ name: 'hourly-sync', queue: 'syncs', cron: '@every 1h', payload: null });

    const rows = sched.list();
    const byName = Object.fromEntries(rows.map((r) => [r.name, r]));
    assert.deepEqual(Object.keys(byName).sort(), ['daily-recap', 'hourly-sync']);
    assert.equal(byName['daily-recap'].queue, 'emails');
    assert.equal(byName['daily-recap'].priority, 3);
    assert.deepEqual(JSON.parse(byName['daily-recap'].payload), { foo: 1 });
    assert.equal(byName['daily-recap'].enabled, true);
    assert.ok(byName['daily-recap'].next_fire_at > 0);
  } finally {
    cleanup();
  }
});

test('pause/resume idempotent', () => {
  const { path: dbPath, open, cleanup } = tmpdb();
  let db;
  try {
    db = open(dbPath);
    const sched = new honker.Scheduler(db);
    sched.add({ name: 'a', queue: 'q', cron: '0 9 * * *', payload: null });

    assert.equal(sched.pause('a'), true);
    assert.equal(sched.pause('a'), false); // already paused
    assert.equal(sched.pause('missing'), false);

    const paused = sched.list().find((s) => s.name === 'a');
    assert.equal(paused.enabled, false);

    assert.equal(sched.resume('a'), true);
    assert.equal(sched.resume('a'), false); // already enabled

    const enabled = sched.list().find((s) => s.name === 'a');
    assert.equal(enabled.enabled, true);
  } finally {
    cleanup();
  }
});

test('update mutates fields and recomputes next_fire_at on cron change', () => {
  const { path: dbPath, open, cleanup } = tmpdb();
  let db;
  try {
    db = open(dbPath);
    const sched = new honker.Scheduler(db);
    sched.add({ name: 't', queue: 'q', cron: '0 9 * * *', payload: { v: 1 }, priority: 0 });

    assert.equal(sched.update('t', { payload: { v: 99 }, priority: 5 }), true);
    const row = sched.list().find((s) => s.name === 't');
    assert.deepEqual(JSON.parse(row.payload), { v: 99 });
    assert.equal(row.priority, 5);

    const before = row.next_fire_at;
    assert.equal(sched.update('t', { cron: '*/5 * * * *' }), true);
    const after = sched.list().find((s) => s.name === 't');
    assert.equal(after.cron_expr, '*/5 * * * *');
    assert.notEqual(after.next_fire_at, before);

    assert.equal(sched.update('missing', { payload: {} }), false);
  } finally {
    cleanup();
  }
});

// ---------- queue cancel / getJob ----------

test('queue.getJob returns row, cancel removes pending', () => {
  const { path: dbPath, open, cleanup } = tmpdb();
  let db;
  try {
    db = open(dbPath);
    const q = db.queue('emails');

    const tx = db.transaction();
    const jid = q.enqueueTx(tx, { to: 'alice@example.com' });
    tx.commit();

    const row = q.getJob(jid);
    assert.equal(row.queue, 'emails');
    assert.equal(row.state, 'pending');
    assert.deepEqual(JSON.parse(row.payload), { to: 'alice@example.com' });
    assert.equal(row.id, jid);

    assert.equal(q.cancel(jid), true);
    assert.equal(q.cancel(jid), false); // idempotent
    assert.equal(q.getJob(jid), null);
    assert.equal(q.claimOne('worker-1'), null);
  } finally {
    cleanup();
  }
});

test('cancel of processing job invalidates ack', () => {
  const { path: dbPath, open, cleanup } = tmpdb();
  let db;
  try {
    db = open(dbPath);
    const q = db.queue('emails');

    const tx = db.transaction();
    const jid = q.enqueueTx(tx, { to: 'x' });
    tx.commit();

    const job = q.claimOne('worker-1');
    assert.equal(job.id, jid);

    assert.equal(q.cancel(jid), true);
    // Worker's ack returns false — same as expired claim.
    assert.equal(job.ack(), false);
  } finally {
    cleanup();
  }
});

File: packages/honker-rs/tests/phase_mantle.rs
//! Phase Mantle tests: schedule lifecycle (pause/resume/list/update)
//! and queue cancel/get_job.

use honker::{Database, EnqueueOpts, QueueOpts, ScheduleUpdate, ScheduledTask};
use serde_json::json;

fn open_db() -> (tempfile::TempDir, Database) {
    let tmp = tempfile::tempdir().unwrap();
    let db = Database::open(tmp.path().join("t.db")).unwrap();
    (tmp, db)
}

#[test]
fn schedule_list_round_trips_all_fields() {
    let (_tmp, db) = open_db();
    let sched = db.scheduler();

    sched
        .add(ScheduledTask {
            name: "daily-recap".into(),
            queue: "emails".into(),
            schedule: "0 9 * * 1".into(),
            payload: json!({"team": "premier-league"}),
            priority: 3,
            expires_s: None,
            max_attempts: None,
        })
        .unwrap();
    sched
        .add(ScheduledTask {
            name: "hourly-sync".into(),
            queue: "syncs".into(),
            schedule: "@every 1h".into(),
            payload: json!(null),
            priority: 0,
            expires_s: None,
            max_attempts: None,
        })
        .unwrap();

    let rows = sched.list().unwrap();
    assert_eq!(rows.len(), 2);
    let recap = rows.iter().find(|r| r.name == "daily-recap").unwrap();
    assert_eq!(recap.queue, "emails");
    assert_eq!(recap.priority, 3);
    assert!(recap.enabled);
    assert!(recap.next_fire_at > 0);
    let payload: serde_json::Value = serde_json::from_str(&recap.payload).unwrap();
    assert_eq!(payload["team"], "premier-league");
}

#[test]
fn pause_resume_idempotent() {
    let (_tmp, db) = open_db();
    let sched = db.scheduler();
    sched
        .add(ScheduledTask {
            name: "a".into(),
            queue: "q".into(),
            schedule: "0 9 * * *".into(),
            payload: json!(null),
            priority: 0,
            expires_s: None,
            max_attempts: None,
        })
        .unwrap();

    assert!(sched.pause("a").unwrap());
    assert!(!sched.pause("a").unwrap()); // idempotent
    assert!(!sched.pause("missing").unwrap());

    let row = sched
        .list()
        .unwrap()
        .into_iter()
        .find(|r| r.name == "a")
        .unwrap();
    assert!(!row.enabled);

    assert!(sched.resume("a").unwrap());
    assert!(!sched.resume("a").unwrap()); // already enabled

    let row = sched
        .list()
        .unwrap()
        .into_iter()
        .find(|r| r.name == "a")
        .unwrap();
    assert!(row.enabled);
}

#[test]
fn update_mutates_fields_and_recomputes_next_fire_at() {
    let (_tmp, db) = open_db();
    let sched = db.scheduler();
    sched
        .add(ScheduledTask {
            name: "t".into(),
            queue: "q".into(),
            schedule: "0 9 * * *".into(),
            payload: json!({"v": 1}),
            priority: 0,
            expires_s: None,
            max_attempts: None,
        })
        .unwrap();

    assert!(
        sched
            .update(
                "t",
                ScheduleUpdate {
                    payload: Some(json!({"v": 99})),
                    priority: Some(5),
                    ..Default::default()
                },
            )
            .unwrap()
    );
    let row = sched
        .list()
        .unwrap()
        .into_iter()
        .find(|r| r.name == "t")
        .unwrap();
    let payload: serde_json::Value = serde_json::from_str(&row.payload).unwrap();
    assert_eq!(payload["v"], 99);
    assert_eq!(row.priority, 5);

    let before = row.next_fire_at;
    assert!(
        sched
            .update(
                "t",
                ScheduleUpdate {
                    cron_expr: Some("*/5 * * * *".into()),
                    ..Default::default()
                },
            )
            .unwrap()
    );
    let row = sched
        .list()
        .unwrap()
        .into_iter()
        .find(|r| r.name == "t")
        .unwrap();
    assert_eq!(row.cron_expr, "*/5 * * * *");
    assert_ne!(row.next_fire_at, before);

    assert!(
        !sched
            .update(
                "missing",
                ScheduleUpdate {
                    payload: Some(json!({})),
                    ..Default::default()
                },
            )
            .unwrap()
    );
}

#[test]
fn update_no_fields_is_noop() {
    let (_tmp, db) = open_db();
    let sched = db.scheduler();
    sched
        .add(ScheduledTask {
            name: "t".into(),
            queue: "q".into(),
            schedule: "0 9 * * *".into(),
            payload: json!({"v": 1}),
            priority: 0,
            expires_s: None,
            max_attempts: None,
        })
        .unwrap();

    let before: Vec<_> = sched
        .list()
        .unwrap()
        .into_iter()
        .map(|r| r.next_fire_at)
        .collect();
    assert!(!sched.update("t", ScheduleUpdate::default()).unwrap());
    let after: Vec<_> = sched
        .list()
        .unwrap()
        .into_iter()
        .map(|r| r.next_fire_at)
        .collect();
    assert_eq!(before, after);
}

#[test]
fn queue_get_job_returns_row_misses_after_cancel() {
    let (_tmp, db) = open_db();
    let q = db.queue("emails", QueueOpts::default());

    let id = q
        .enqueue(&json!({"to": "alice@example.com"}), EnqueueOpts::default())
        .unwrap();

    let row = q.get_job(id).unwrap().expect("should hit");
    assert_eq!(row.id, id);
    assert_eq!(row.queue, "emails");
    assert_eq!(row.state, "pending");

    assert!(q.cancel(id).unwrap());
    assert!(!q.cancel(id).unwrap()); // idempotent
    assert!(q.get_job(id).unwrap().is_none());
    assert!(q.claim_one("worker-1").unwrap().is_none());
}

#[test]
fn cancel_processing_invalidates_ack() {
    let (_tmp, db) = open_db();
    let q = db.queue("emails", QueueOpts::default());
    let id = q
        .enqueue(&json!({"to": "x"}), EnqueueOpts::default())
        .unwrap();
    let job = q.claim_one("worker-1").unwrap().unwrap();
    assert_eq!(job.id, id);

    assert!(q.cancel(id).unwrap());
    // Worker's ack now returns false (same shape as expired claim).
    assert!(!job.ack().unwrap());
}

#[test]
fn paused_schedule_does_not_emit_on_tick() {
    let (_tmp, db) = open_db();
    let sched = db.scheduler();
    sched
        .add(ScheduledTask {
            name: "due".into(),
            queue: "emails".into(),
            schedule: "@every 1s".into(),
            payload: json!({"x": 1}),
            priority: 0,
            expires_s: None,
            max_attempts: None,
        })
        .unwrap();
    std::thread::sleep(std::time::Duration::from_millis(1100));
    sched.pause("due").unwrap();

    let fires = sched.tick().unwrap();
    assert_eq!(
        fires.len(),
        0,
        "paused schedule must not emit; got {fires:?}"
    );

    // Resume and tick again — now it fires.
    sched.resume("due").unwrap();
    let fires2 = sched.tick().unwrap();
    assert!(
        !fires2.is_empty(),
        "resumed schedule should emit; got {fires2:?}"
    );
}

#[test]
fn queue_get_job_misses_after_ack() {
    let (_tmp, db) = open_db();
    let q = db.queue("emails", QueueOpts::default());
    let id = q
        .enqueue(&json!({"to": "x"}), EnqueueOpts::default())
        .unwrap();

    let job = q.claim_one("worker-1").unwrap().unwrap();
    assert_eq!(job.id, id);
    assert!(job.ack().unwrap());
    // Row gone after ack — get_job misses just like after cancel.
    assert!(q.get_job(id).unwrap().is_none());
}

#[test]
fn update_payload_null_vs_omitted() {
    let (_tmp, db) = open_db();
    let sched = db.scheduler();
    sched
        .add(ScheduledTask {
            name: "t".into(),
            queue: "q".into(),
            schedule: "0 9 * * *".into(),
            payload: json!({"v": 1}),
            priority: 0,
            expires_s: None,
            max_attempts: None,
        })
        .unwrap();

    // Omitted payload — payload field stays as Default (None).
    sched
        .update(
            "t",
            ScheduleUpdate {
                priority: Some(7),
                ..Default::default()
            },
        )
        .unwrap();
    let row = sched
        .list()
        .unwrap()
        .into_iter()
        .find(|r| r.name == "t")
        .unwrap();
    let payload: serde_json::Value = serde_json::from_str(&row.payload).unwrap();
    assert_eq!(payload, json!({"v": 1}));
    assert_eq!(row.priority, 7);

    // payload: Some(Value::Null) — explicitly write JSON null.
    sched
        .update(
            "t",
            ScheduleUpdate {
                payload: Some(serde_json::Value::Null),
                ..Default::default()
            },
        )
        .unwrap();
    let row = sched
        .list()
        .unwrap()
        .into_iter()
        .find(|r| r.name == "t")
        .unwrap();
    let payload: serde_json::Value = serde_json::from_str(&row.payload).unwrap();
    assert!(payload.is_null());
}
